homeassistant/components/rachio/binary_sensor.py

```diff
@@ -43,6 +46,68 @@
 _LOGGER = logging.getLogger(__name__)
 
 
+@dataclass(frozen=True, kw_only=True)
+class RachioControllerBinarySensorDescription(BinarySensorEntityDescription):
+    """Describe a Rachio controller binary sensor."""
+
+    update_received: Callable[[str], bool | None]
+    is_on: Callable[[RachioIro], bool]
+    signal_string: str
+
+
+CONTROLLER_BINARY_SENSOR_TYPES: tuple[RachioControllerBinarySensorDescription, ...] = (
+    RachioControllerBinarySensorDescription(
+        key=KEY_ONLINE,
+        device_class=BinarySensorDeviceClass.CONNECTIVITY,
+        signal_string=SIGNAL_RACHIO_CONTROLLER_UPDATE,
+        is_on=lambda controller: controller.init_data[KEY_STATUS] == STATUS_ONLINE,
+        update_received=lambda state: True
+        if state in (SUBTYPE_ONLINE, SUBTYPE_COLD_REBOOT)
+        else False
+        if state == SUBTYPE_OFFLINE
+        else None,
+    ),
+    RachioControllerBinarySensorDescription(
+        key=KEY_RAIN_SENSOR,
+        translation_key="rain",
+        device_class=BinarySensorDeviceClass.MOISTURE,
+        signal_string=SIGNAL_RACHIO_RAIN_SENSOR_UPDATE,
+        is_on=lambda controller: controller.init_data[KEY_RAIN_SENSOR_TRIPPED],
+        update_received=lambda state: True
+        if state == SUBTYPE_RAIN_SENSOR_DETECTION_ON
+        else False
+        if state == SUBTYPE_RAIN_SENSOR_DETECTION_OFF
+        else None,
+    ),
```

Comment:
idem

Reply:
this one might be a dict lookup

Reply:
I just left as is, it seemed possible as a dict lookup but then I had issues with MyPy complaining about unnecessary lambdas and wanted a named function.

Reply:
```py
        update_received={
            SUBTYPE_RAIN_SENSOR_DETECTION_ON: True,
            SUBTYPE_RAIN_SENSOR_DETECTION_OFF: False,
        }.get,
```

Reply:
I was overthinking it. Will change. Thx